services/cargohold/test/integration/API/V4.hs

```diff
@@ -0,0 +1,282 @@
+          <!! const 201 === statusCode
+      let loc = decodeHeader "Location" r1 :: ByteString
+      let Just ast = responseJsonMaybe @Asset r1
+      let Just tok = view assetToken ast
+      -- Check mandatory Date header
+      let Just date = C8.unpack <$> lookup "Date" (responseHeaders r1)
+      let utc = parseTimeOrError False defaultTimeLocale rfc822DateFormat date :: UTCTime
+      -- Potentially check for the expires header
+      when (isJust $ join (assetRetentionSeconds <$> (sets ^. setAssetRetention))) $ do
+        liftIO $ assertBool "invalid expiration" (Just utc < view assetExpires ast)
+      -- Lookup with token and download via redirect.
+      r2 <-
+        get (c . path loc . zUser uid . header "Asset-Token" (toByteString' tok) . noRedirect) <!! do
+          const 302 === statusCode
+          const Nothing === responseBody
+      r3 <- flip get' id =<< parseUrlThrow (C8.unpack (getHeader' "Location" r2))
+      liftIO $ do
+        assertEqual "status" status200 (responseStatus r3)
+        assertEqual "content-type should always be application/octet-stream" (Just applicationOctetStream) (getContentType r3)
+        assertEqual "token mismatch" tok (decodeHeader "x-amz-meta-token" r3)
+        assertEqual "user mismatch" uid (decodeHeader "x-amz-meta-user" r3)
+        assertEqual "data mismatch" (Just "Hello World") (responseBody r3)
+      -- Delete (forbidden for other users)
+      deleteAsset c uid2 (view assetKey ast) !!! const 403 === statusCode
+      -- Delete (allowed for creator)
+      deleteAsset c uid (view assetKey ast) !!! const 200 === statusCode
+      r4 <-
+        get (c . path loc . zUser uid . header "Asset-Token" (toByteString' tok) . noRedirect)
+          <!! const 404 === statusCode
+      let Just date' = C8.unpack <$> lookup "Date" (responseHeaders r4)
+      let utc' = parseTimeOrError False defaultTimeLocale rfc822DateFormat date' :: UTCTime
+      liftIO $ assertBool "bad date" (utc' >= utc)
+
+testSimpleTokens :: TestSignature ()
+testSimpleTokens c = do
+  uid <- liftIO $ Id <$> nextRandom
+  uid2 <- liftIO $ Id <$> nextRandom
+  -- Initial upload
+  let sets = defAssetSettings & set setAssetRetention (Just AssetVolatile)
+  let bdy = (applicationText, "Hello World")
+  r1 <-
+    uploadSimple (c . path "/assets/v3") uid sets bdy
+      <!! const 201 === statusCode
+  let loc = decodeHeader "Location" r1 :: ByteString
+  let Just ast = responseJsonMaybe @Asset r1
+  let key = view assetKey ast
+  let Just tok = view assetToken ast
+  -- No access without token from other user (opaque 404)
+  get (c . path loc . zUser uid2 . noRedirect)
+    !!! const 404 === statusCode
+  -- No access with wrong token (opaque 404)
+  get (c . path loc . zUser uid2 . header "Asset-Token" "acb123" . noRedirect)
+    !!! const 404 === statusCode
+  -- Token renewal fails if not done by owner
+  post (c . paths ["assets", "v3", toByteString' (qUnqualified key), "token"] . zUser uid2) !!! do
```

Comment:
Isn't there a utility function for making this request? And should it be v3 or v4?

Reply:
It's a bit unfortunate, but this utility function will be introduced and used with the next PR: https://github.com/wireapp/wire-server/pull/2004/files#diff-e9ecc1bf7fe0d5557306c1ab67374434916fe268af016613313a503dc19f2cd5R146

@mdimjasevic , would it be okay to keep this for now in this PR? As the proposed change is already available in the upcoming / next PR...

Reply:
Sure, introducing it in the next PR sounds good!